prov/util/src/util_cq.c

```diff
@@ -203,21 +203,35 @@ ssize_t ofi_cq_readerr(struct fid_cq *cq_fid, struct fi_cq_err_entry *buf,
 	struct util_cq *cq;
 	struct util_cq_err_entry *err;
 	struct slist_entry *entry;
-	ssize_t ret;
+	char *err_buf_save = NULL;
+	size_t err_data_size = 0;
+	uint32_t api_version;
+	ssize_t ret = -FI_EAGAIN;
 
 	cq = container_of(cq_fid, struct util_cq, cq_fid);
+	api_version = cq->domain->fabric->fabric_fid.api_version;
+
 	fastlock_acquire(&cq->cq_lock);
-	if (!ofi_cirque_isempty(cq->cirq) &&
-	    (ofi_cirque_head(cq->cirq)->flags & UTIL_FLAG_ERROR)) {
-		ofi_cirque_discard(cq->cirq);
-		entry = slist_remove_head(&cq->err_list);
-		err = container_of(entry, struct util_cq_err_entry, list_entry);
-		*buf = err->err_entry;
-		free(err);
-		ret = 0;
-	} else {
-		ret = -FI_EAGAIN;
+	if (ofi_cirque_isempty(cq->cirq) ||
+	    !(ofi_cirque_head(cq->cirq)->flags & UTIL_FLAG_ERROR))
+		goto unlock;
+
+	ofi_cirque_discard(cq->cirq);
+	entry = slist_remove_head(&cq->err_list);
+	err = container_of(entry, struct util_cq_err_entry, list_entry);
+	if ((FI_VERSION_GE(api_version, FI_VERSION(1, 5))) && buf->err_data_size) {
+		err_data_size = MIN(buf->err_data_size, err->err_entry.err_data_size);
+		memcpy(buf->err_data, err->err_entry.err_data, err_data_size);
+		err_buf_save = buf->err_data;
+	}
+	*buf = err->err_entry;
+	if (err_buf_save) {
+		buf->err_data = err_buf_save;
+		buf->err_data_size = err_data_size;
```

Comment:
The above would be clearer if the two if statements were merged.  Something like:
````
if (api_version...) {
    ...
    save_err_data = buf->err_data;
    *buf = err->err_entry;
    *buf->err_data = save_err_data;
} else {
    *buf = err->err_entry;
}
````
The second if statement only executes if the first if statement was taken.  The only reason for separating things is to have a single common *buf assignment, but at the cost of introducing a second branch.